state-chain/pallets/cf-trading-strategy/src/lib.rs

```diff
@@ -0,0 +1,354 @@
+		#[pallet::call_index(1)]
+		#[pallet::weight(Weight::zero())] // TODO: benchmark
+		pub fn deploy_trading_strategy(
+			origin: OriginFor<T>,
+			base_asset_amount: AssetAmount,
+			quote_asset_amount: AssetAmount,
+			base_asset: Asset,
+			strategy: TradingStrategy,
+		) -> DispatchResult {
+			let lp = &T::AccountRoleRegistry::ensure_liquidity_provider(origin)?;
+
+			let quote_asset = STABLE_ASSET;
+
+			let strategy_id = {
+				// Check that strategy is created with sufficient funds:
+				{
+					let fraction_of_required = |asset, provided| {
+						let min_required = MinimumDeploymentAmountForStrategy::<T>::get(asset);
+
+						if min_required == 0 {
+							FixedU64::one()
+						} else {
+							FixedU64::from_rational(provided, min_required)
+						}
+					};
+
+					ensure!(
+						fraction_of_required(base_asset, base_asset_amount) +
+							fraction_of_required(quote_asset, quote_asset_amount) >=
+							FixedU64::one(),
+						Error::<T>::AmountBelowDeploymentThreshold
+					);
+				}
+
+				let nonce = frame_system::Pallet::<T>::account_nonce(lp);
+
+				let strategy_id = derive_strategy_id::<T>(lp, nonce);
+
+				Self::deposit_event(Event::<T>::StrategyDeployed {
+					account_id: lp.clone(),
+					strategy_id: strategy_id.clone(),
+					base_asset,
+					strategy: strategy.clone(),
+				});
+
+				Strategies::<T>::insert(
+					strategy_id.clone(),
+					TradingStrategyEntry { base_asset, owner: lp.clone(), strategy },
+				);
+
+				strategy_id
+			};
+
+			Self::add_funds_to_existing_strategy(
+				lp,
+				&strategy_id,
+				base_asset,
+				base_asset_amount,
+				quote_asset_amount,
+			)
+		}
+
+		#[pallet::call_index(2)]
+		#[pallet::weight(Weight::zero())] // TODO: benchmark
+		pub fn close_strategy(origin: OriginFor<T>, strategy_id: T::AccountId) -> DispatchResult {
+			let lp = &T::AccountRoleRegistry::ensure_liquidity_provider(origin)?;
+
+			let TradingStrategyEntry { base_asset, owner, strategy } =
+				Strategies::<T>::take(&strategy_id).ok_or(Error::<T>::StrategyNotFound)?;
+
+			ensure!(lp == &owner, Error::<T>::InvalidOwner);
+
+			// TODO: instead of reading ticks from the strategy, we could extend PoolApi with
+			// a method to close all (limit) orders (which might be necessary for more complex
+			// strategies).
+			let TradingStrategy::SellAndBuyAtTicks { buy_tick, sell_tick } = strategy;
+
+			let cancel_limit_orders = |side, tick| {
+				// TODO: check if order cancellation is infallible?
+				T::PoolApi::cancel_limit_order(
+					&strategy_id,
+					base_asset,
+					STABLE_ASSET,
+					side,
+					STRATEGY_ORDER_ID,
+					tick,
+				)
+			};
+
+			cancel_limit_orders(Side::Buy, buy_tick)?;
+			cancel_limit_orders(Side::Sell, sell_tick)?;
+
+			for asset in [base_asset, STABLE_ASSET] {
+				let balance = T::BalanceApi::get_balance(&strategy_id, asset);
+				T::BalanceApi::try_debit_account(&strategy_id, asset, balance)?;
+				T::BalanceApi::credit_account(lp, asset, balance);
```

Comment:
Probably makes sense to have a `transfer` function on the balance API? so we don't need to pair these every time - (that function would just pair them like this)